[WIP] Hard error when ignoring tensors.

Narsil opened the pull request:
# What does this PR do?

 Better selection/error when saving a checkpoint.

- Find all names we should normally drop (those are in the transformers
  config)
- Find all disjoint tensors (for those we can safely trigger a copy to
  get rid of the sharing before saving)
- Clone those disjoint tensors getting rid of the issue
- Find all identical names (those should be declared in the config
  but we try to find them all anyway.)
- For all identical names:
  - If they are in the config, just ignore them everything is fine
  - If they are not, warn about them.
- For all remainder tensors which are shared yet neither identical NOR
  disjoint. raise a hard error.

* Adding a failing test on `main` that passes here.

* We don't need to keep the subfolder logic in this test.


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Narsil committed:
[WIP] Hard error when ignoring tensors.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27484). All of your documentation changes will be reflected on that endpoint.

Narsil committed:
Better selection/error when saving a checkpoint.

- Find all names we should normally drop (those are in the transformers
  config)
- Find all disjoint tensors (for those we can safely trigger a copy to
  get rid of the sharing before saving)
- Clone those disjoint tensors getting rid of the issue
- Find all identical names (those should be declared in the config
  but we try to find them all anyway.)
- For all identical names:
  - If they are in the config, just ignore them everything is fine
  - If they are not, warn about them.
- For all remainder tensors which are shared yet neither identical NOR
  disjoint. raise a hard error.

Narsil committed:
Adding a failing test on `main` that passes here.

Narsil committed:
We don't need to keep the subfolder logic in this test.

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -515,6 +515,64 @@ def set_initialized_submodules(model, state_dict_keys):
+
+        _, last_stop, last_name = areas[0]
+        filtered_tensors.append({last_name})
+        for start, stop, name in areas[1:]:
+            if start >= last_stop:
+                filtered_tensors.append({name})
+            else:
+                filtered_tensors[-1].add(name)
```
here we merge the tensors that have the same ending in the filtered tensors

Narsil replied:
merged the tensors ?

We add the name to a previous set, so that the list `filtered_tensors` will contain a set with at least 2 entries.

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -515,6 +515,64 @@ def set_initialized_submodules(model, state_dict_keys):
+    return stop
+
+
+def _find_disjoint(tensors: List[Set[str]], state_dict: Dict[str, torch.Tensor]) -> Tuple[List[Set[str]], Set[str]]:
```
this is tested

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -515,6 +515,64 @@ def set_initialized_submodules(model, state_dict_keys):
+    return shared_tensors, disjoint_tensors
+
+
+def _find_identical(tensors: List[Set[str]], state_dict: Dict[str, torch.Tensor]) -> Tuple[List[Set[str]], Set[str]]:
```
This does not seem tested by the small dummy test (but is most probably tested by some models that do have this?)

Narsil replied:
Yes, because `area` contains the device as a key.

Nothing in transformers could trigger this, only `optimum` did trigger this. It requires some low level manipulation of storages and tensors.

Narsil replied:
Identical was already previously tested by other tests. (Since it's a regular flow for shared tensors with `tie_weights`.)

The dummy test only tests the new path for disjoint tensors (which cannot be triggered by regular transformers code)

ArthurZucker reviewed (commented):
Left a few questions, it's very core so leaving some doc will help other futur maintainers!

Narsil commented on src/transformers/modeling_utils.py:
```diff
@@ -2121,6 +2179,8 @@ def save_pretrained(
             # These are all the pointers of shared tensors.
             shared_ptrs = {ptr: names for ptr, names in ptrs.items() if len(names) > 1}
             warn_names = set()
+            error_names = set()
```
It's just a sort of code comment, it's mostly that some names will trigger warn, some errors.

Not strongly feeling here, just feels it's a bit clearer that there are different names.

Narsil commented on src/transformers/modeling_utils.py:
```diff
@@ -2131,25 +2191,41 @@ def save_pretrained(
+            # the config, better show a proper warning.
+            names, identical_names = _find_identical(names, state_dict)
+            for inames in identical_names:
+                known = inames.intersection(to_delete_names)
```
Makes every bit less entangled. `_find_xx` don't care about transformers specifications on what tensors are tied.
`to_delete_names` IS the only place where we know tensors we should delete.

Set calculations should be ridiculously fast compared to the actual saving/loading part.

Narsil commented on src/transformers/modeling_utils.py:
```diff
@@ -2131,25 +2191,41 @@ def save_pretrained(
+                    del state_dict[name]
+                    warn_names.add(name)
+
+            error_names.update(names)
```
```suggestion
            error_names.update(shared_names)
```

Narsil committed:
Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ArthurZucker reviewed (approved):
Thanks! I would add 1 or two comments to the private functions for posterity but impressive otherwise 👍🏻

fxmarty commented:
Removing the `.clone()` [here](https://github.com/huggingface/optimum/blob/15a162824d0c5d8aa7a3d14ab6e9bb07e5732fb6/optimum/bettertransformer/models/base.py#L154), I can confirm that

```python
from transformers import AutoModel
from optimum.bettertransformer import BetterTransformer
import torch

model = AutoModel.from_pretrained("bert-base-uncased")

model_bt = BetterTransformer.transform(model)

model_reverse = BetterTransformer.reverse(model_bt)

model_reverse.save_pretrained("bt-reverse-saved")

model = AutoModel.from_pretrained("bert-base-uncased")
model_reverse = AutoModel.from_pretrained("bt-reverse-saved")

assert torch.equal(model.encoder.layer[0].attention.self.query.weight, model_reverse.encoder.layer[0].attention.self.query.weight)
assert torch.equal(model.encoder.layer[0].attention.self.key.weight, model_reverse.encoder.layer[0].attention.self.key.weight)
assert torch.equal(model.encoder.layer[0].attention.self.value.weight, model_reverse.encoder.layer[0].attention.self.value.weight)
```

works fine with this fix (while was failing on main).

fxmarty commented on src/transformers/modeling_utils.py:
```diff
@@ -2131,25 +2192,42 @@ def save_pretrained(
+            # We are entering a place where the weights and the transformers configuration do NOT match.
+            shared_names, disjoint_names = _find_disjoint(shared_ptrs.values(), state_dict)
+            # Those are actually tensor sharing but disjoint from each other, we can safely clone them
+            # Reloaded won't have the same property, but it shouldn't matter in any meaningful way.
```
What property means in this context?

fxmarty reviewed (approved):
LGTM

ArthurZucker committed:
Merge branch 'main' into raise_hard_error

ArthurZucker commented:
Forgot to merge after resolving conflicts with main. Merging now
